### DISCUSSION TITLE
Implement Every Interface

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#2801] [Language Ideas]
It would be nice if I could do the following:
````
public class Foo : IEnumerable<T> {
   public IEnumerator<T> GetEnumerator<T>(){
      //....
   }
}
````
With something like this, GetEnumerator would be implemented for every T.

**Comment by CyrusNajmabadi** [comment]
I don't know what this is asking for.

**Comment by franchyd** [comment]
I think he wants a non-generic type to be enumerable as any type ._.  I can't fathom as to why though.

**Comment by Pretasoc** [comment]
> With something like this, GetEnumerator would be implemented for every T.

Looks like Extension everithing or shapes to me

**Comment by CyrusNajmabadi** [comment]
> I think he wants a non-generic type to be enumerable as any type

Ah, got it.  I'm not sure what this means at all either.  :)

**Comment by yaakov-h** [comment]
If I'm interpreting this correctly...

Currently a generic interface requires the implementing class either be generic itself (`class Foo<T> : IBar<T>`, or implement a specific variant (`class Foo<Baz> : IBar<Baz>`).

What you're suggesting is that a non-generic class can implement a generic interface for any `T` by providing generic methods that match those defined on the interface.

So in your example, I could cast `Foo` to `IEnumerable<int>` and `GetEnumerator<int>` would be called, but if I cast it to `IEnumerable<string>` then `GetEnumerator<string>` would be called.

Is that correct?

**Comment by TonyValenti** [comment]
@yaakov-h Yes!  That is correct!

**Comment by CyrusNajmabadi** [comment]
could you do it like so:

```c#
class WhateverFactory
{
  public static IEnumerable<T> Make<T>() => new Whatever<T>();

  private class Whatever<T> : IEnumerable<T> { ... your impl ... } 
}
```

Also, @TonyValenti can you explain the use case where you want this?  The request is too isolated from a real use case, so i don't know why you'd want this.

**Comment by CyrusNajmabadi** [comment]
> @yaakov-h Yes! That is correct!

So how would that actually work? How will you actually provide the stream of integers if someone casts to `IEnumerable<int>` or the stream of strings if they cast to `IEnumerable<string>`?

Honestly curious here.  Could you show what your implementation of `GetEnumerator` would be in your original post?

**Comment by brabebhin** [comment]
Arguably, this could be useful when dealing with numeric collections, which can be converted rather easily in between types (int to double to uint to string to etc).

**Comment by franchyd** [comment]
@mcosmin222 That sounds more like #110 #1711 or #2038 .  A completely unconstrained setup would not lend well to assuming its a numeric type.

**Comment by brabebhin** [comment]
Obviously op did not clearly state their purposes, but imo this is what they want. It is obviously reasonable to only do the cast between types which are convertible - maybe by overriding the explicit conversion operator.
